kadai1/uesho/imgconv/imgconv.go

```diff
@@ -0,0 +1,176 @@
+package imgconv
+
+import (
+	"fmt"
+	"image"
+	"image/gif"
+	"image/jpeg"
+	"image/png"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/image/bmp"
+)
+
+type ImageExt string // 画像ファイルの拡張子
+
+type imageConverter struct {
+	from ImageExt
+	to   ImageExt
+}
+
+const (
+	GIF  ImageExt = "gif"
+	JPEG ImageExt = "jpeg"
+	JPG  ImageExt = "jpg"
+	PNG  ImageExt = "png"
+	BMP  ImageExt = "bmp"
+)
+
+var ValidImageExts = []ImageExt{GIF, JPEG, JPG, PNG, BMP}
+
+// 画像を変換する
+func Do(args map[string]string) error {
+	converter, err := newImageConverter(args["from"], args["to"])
+	if err != nil {
+		return err
+	}
+
+	path := args["dir"]
+
+	err = converter.convertAll(path)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// String型にする
+func (ext ImageExt) toString() string {
+	return string(ext)
+}
+
+// ImageExt型にする
+func toImageExt(str string) (*ImageExt, error) {
+	if str[0] == '.' {
+		str = str[1:]
+	}
+
+	for _, v := range ValidImageExts {
+		if v.toString() == strings.ToLower(str) {
+			return &v, nil
+		}
+	}
+
+	return nil, fmt.Errorf("拡張子が正しくありません: %s", str)
+}
+
+// imageConverter型を作成する
+func newImageConverter(from, to string) (*imageConverter, error) {
+	extFrom, err := toImageExt(from)
+	if err != nil {
+		return nil, err
+	}
+
+	extTo, err := toImageExt(to)
+	if err != nil {
+		return nil, err
+	}
+
+	return &imageConverter{from: *extFrom, to: *extTo}, nil
+}
+
+// 画像を読み込む
+func (image_converter imageConverter) readImage(path string) (image.Image, error) {
```

Comment:
```suggestion
func (c *imageConverter) readImage(path string) (image.Image, error) {
```

- Goでは変数名をcamelCaseにします。また、レシーバには短い名前を使うのが一般的です。
- この場合は（構造体のコピーは必要ないので）ポインタレシーバで良いと思います。https://golang.org/doc/effective_go.html#pointers_vs_values

Reply:
一文字で良いんですね。Effective Go 一通り読んでいきたいと思います！